Product/Production/Common/CONNECTCoreLib/src/main/java/gov/hhs/fha/nhinc/callback/cxf/CertificateChainValidator.java

```diff
@@ -0,0 +1,213 @@
+    private final CertificateManager certificateManager;
+
+    // http://docs.oracle.com/javase/7/docs/api/java/security/cert/X509Extension.html#getExtensionValue(java.lang.String)
+    private static final String AUTHORITY_KEY_ID = "2.5.29.35";
+    private static final String SUBJECT_KEY_ID = "2.5.29.14";
+
+    /**
+     *
+     */
+    private CertificateChainValidator() {
+        certificateManager = CertificateManagerImpl.getInstance();
+    }
+
+    /**
+     * @param certificateManager
+     */
+    CertificateChainValidator(CertificateManager certificateManager) {
+        this.certificateManager = certificateManager;
+    }
+    public static CertificateChainValidator getInstance() {
+        return new CertificateChainValidator();
+    }
+
+    /**
+     * ALL root CA are self-sign cert. Verify if cert is self-sign or chain of trust
+     *
+     * @param certToCheck
+     * @return true if self-sign cert or false if chain of trust
+     */
+    public boolean isPartChainOfTrust(X509Certificate certToCheck) {
+        try {
+            // Try to verify certificate signature with its own public key
+            PublicKey key = certToCheck.getPublicKey();
+            certToCheck.verify(key);
+            return false;
+        } catch (SignatureException | InvalidKeyException | CertificateException | NoSuchAlgorithmException
+            | NoSuchProviderException sigEx) {
+            // LOG.warn(sigEx.getLocalizedMessage(),sigEx);
+            return true;
+        }
+
+    }
+
+    /**
+     * Validate chain of trust.
+     *
+     * @param certToCheck
+     * @return
+     */
+    public boolean validateCert(X509Certificate certToCheck) {
+        List<X509Certificate> allCerts = getAllCertInKeyStore(certificateManager.getTrustStore());
+        // check if cert is chain of trust
+        if (isPartChainOfTrust(certToCheck)) {
+            // extract intermediate/root out and compare with trust store
+            // extract intermediate
+            try {
+                // ---------
+                LOG.debug("Cert serial number to verify against our trustStore {}", certToCheck.getSerialNumber()
+                    .toString(16));
+                String leafIssuerString = certToCheck.getIssuerX500Principal().getName();
+                LOG.debug("Issuer for parent {} ", leafIssuerString);
+
+                String authorityKeyIdentifier = getAuthorityKeyIdentify(certToCheck.getExtensionValue(AUTHORITY_KEY_ID));
+                for (X509Certificate cert : allCerts) {
+                    String name = cert.getSubjectX500Principal().getName();
+                    if (leafIssuerString.equalsIgnoreCase(name)) {
+                        String subjectKeyId = getSubjectKeyID(cert.getExtensionValue(SUBJECT_KEY_ID));
+                        if (StringUtils.isNotBlank(subjectKeyId) && StringUtils.isNotBlank(authorityKeyIdentifier)
+                            && subjectKeyId.equalsIgnoreCase(authorityKeyIdentifier)) {
```

Comment:
We have recursion method after finding the match.  This operation will allow us to go deep more than 2 level.